### PR TITLE
Add optional POPSpa post-training sparsification

### PR DESCRIPTION
**Pull request by lyehe**
## Summary

- Add POPSpa as an opt-in sparsity method alongside the existing, unchanged Opacity ADMM default.
- Implement contribution scoring, intermediate exact pruning, sparse optimization, final exact pruning, and photometric recovery using native C++/CUDA with FastGS and the bundled gsplat/GUT backend. No Python gsplat dependency is added.
- Expose the method and its settings through the Training panel, CLI, JSON, and Python, with matching keys in all ten shipped locales and README usage instructions.
- Preserve frozen Gaussians, validate budgets before topology changes, and include POPSpa state in checkpoint resume and training snapshots.

The optimization profile follows the [pinned KissGsVanilla reference](https://github.com/OpsiClear/KissGsVanilla/tree/5d5212a33a45a6668beb822f17b4b2d88e0f5f8a). This is an independent native implementation of that repository's unofficial POPSpa profile, not the authors' unreleased encoder or a claim of bit-identical PyTorch behavior.

## Design and supporting fixes

- A dedicated controller owns the post-training phases; the trainer retains rendering, camera, and lock ownership. Compaction reuses existing strategy checkpoint adoption instead of adding a general-purpose sparsity interface.
- Checkpoint format advances to v4 for POPSpa state. Reading v1-v3 remains supported; newly written v4 checkpoints require an updated application.
- Zero tensor-header padding for deterministic serialized state without changing the existing tensor file layout.
- Fence reuse of the standalone FastGS Adam batch table to fix an asynchronous upload/kernel race exposed by exact checkpoint-resume regression tests.
- Share the new tests between the full test suite and a focused `popspa_tests` target. Make an existing dataset-path assertion portable to Windows and keep parameter golden files LF-normalized.

## Viewport freeze follow-up

- Bound finite streamless arena acquisitions across the GPU handoff as well as CPU ownership. Pending GPU work completes on a worker while the arena remains reserved; completed producer events avoid waiting for unrelated CUDA streams.
- Invalidate non-owning Vulkan scratch aliases when their owners retire, preventing later projection from reusing dead buffer handles. Existing timeline retirement and external-buffer ownership are preserved.
- Add regression coverage for pending handoffs, unrelated streams, move/reset lifetime, and scratch aliases. These are shared CUDA/Vulkan integration fixes; POPSpa's optimization algorithm is unchanged.

See [investigation and validation details](https://github.com/lyehe/LichtFeld-Studio/blob/feat/popspa-sparsification/eval/popspa-ui-freeze-investigation.md). A controlled 15 ms acquisition previously blocked for about 1,009 ms; it now defers in about 15.2 ms. The reporter's exact freeze has not been reproduced, so this follow-up does not establish that every reported freeze is resolved.

## Latest follow-up validation

- Windows Release application, dependent DLLs, and Python extension rebuilt together (MSVC 19.44, CUDA 13.3, RTX 5090).
- Focused local runners using repository test sources: 21 arena/shared-scratch tests and 8 buffer-retirement tests passed; 9 Python output-lifetime tests passed.
- Garden GUI trials on FastGS and GUT: 2.56M to 512k Gaussians, two scoring passes over 161 cameras, shortened 10-step optimization and 10-step refinement phases. With camera movement and Khronos validation active, neither final trial recorded validation errors, device loss, capture errors, or window-message timeouts.
- Formatting, Windows source preflight, error-debt audit, and staged diff checks passed.

The local validation checkout includes the previously reported COLMAP calibration correction, which is outside this follow-up commit. The full repository test suite was not run. These short GUI trials are responsiveness checks, not quality benchmarks.

## Earlier feature validation

- Windows Release build: `LichtFeld-Studio`, its Python extension, and `popspa_tests` passed (MSVC 19.44, CUDA 13.3, RTX 5090).
- `ctest --test-dir build -C Release -R '^POPSpaContracts$' --output-on-failure`: 66 native tests passed, with no exclusions.
- Python property-view, localization, and training-panel regression suites: 101 tests passed, with no exclusions.
- `clang-format --dry-run --Werror` on changed/new C++ and CUDA files: passed.
- `python -X utf8 tools/check_locale_completeness.py`: passed.
- `python -X utf8 tools/check_ui_hardcoded.py`: passed.
- `git diff --check`: passed.

These original feature results are historical, from the feature branch based on `1cc94ba8`; they were not rerun for the viewport-only follow-up.

The focused native coverage includes scoring on both backends, regularization, exact compaction and frozen-row handling, controller transitions, pipeline execution, parameter contracts, snapshots, and checkpoint compatibility/resume. Subsequent validation and benchmark results are recorded in the PR discussion; the original focused results above are not a claim that the complete repository test suite passes.

## Try it

In the Training panel, enable **Sparsity**, choose **POPSpa** under **Sparsity Settings â†’ Sparsity Method**, and set **Target Gaussians**. The equivalent CLI options are:

```text
--enable-sparsity --sparsity-method popspa --popspa-target-count 256000
```

### CODE REVIEW

File: src/training/trainer_popspa.cpp [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+            cameras.reserve(train_dataset_->size());
+            // Use the dataset's split indices: get_cameras() also contains held-out views.
+            for (size_t i = 0; i < train_dataset_->size(); ++i) {
+                auto* cam = train_dataset_->get_camera(i);
+                if (!cam->image_size_loaded())
+                    cam->load_image_size(params_.dataset.resize_factor, params_.dataset.max_width);
+                cameras.push_back(cam);
+            }
+            const uint64_t fingerprint = camera_fingerprint(cameras);
+            std::sort(cameras.begin(), cameras.end(), [](const auto* a, const auto* b) { return a->uid() < b->uid(); });
```
[the remark this answers is not in the document]
Comment (lyehe) [reply src/training/trainer_popspa.cpp]: Fixed in b495e4ee. The trainer now canonicalizes the camera list by UID before computing its fingerprint, and uses that same list for scoring. Added regression coverage for all six permutations of three cameras across a serialized/restored partial score pass, including preservation of the next score view. Negative tests still detect changed camera membership, poses, and intrinsics. Validation: Windows Release app/extension/test build passed; all 68 native POPSpa tests and 101 Python/UI tests passed. Also addressed the error-debt CI findings by documenting the typed-error conversion boundaries using the audit's supported annotations; the unchanged baseline gate passes locally.

**Comment by lyehe** [comment]
Merged upstream master through fe7d9029 in 3d3d5991, preserving the maintainer's POPSpa completion-evaluation and snapshot fixes. The final-evaluation conflict now retains both POPSpa final-model evaluation and ordinary terminal-resume scheduled evaluation. Follow-up 770154fa adds the shared checkpoint test helper required by the merged tests to popspa_tests. Validation: Windows Release app, extension, and focused test build passed; 72 native tests passed with 2 bicycle-fixture-dependent snapshot tests skipped; 101 Python/UI tests and 36 Windows preflight tests passed. Error-debt, localization, UI-text, and Windows source-preflight audits also passed. CI should rerun on the updated head.

**Comment by MrNeRF** [comment]
Built and tested head 770154fa on Linux (RTX 4080, CUDA 13.3, GCC 14), then ran the Mip-NeRF 360 benchmark with the official MRNF config (30k iterations, 1M cap, `--eval --test-every 8`). Same binary for all runs.

**Benchmark, mean over 7 scenes**

| | PSNR | SSIM | LPIPS | Gaussians | time / scene |
|---|---:|---:|---:|---:|---:|
| regular training | 29.27 | 0.876 | 0.178 | 1,000,000 | 4.5 min |
| POPSpa (target 256k, defaults) | 28.70 | 0.851 | 0.229 | 256,000 | 5.5 min |
| Opacity ADMM (defaults, 60% pruned) | 28.88 | 0.853 | 0.225 | 400,000 | 5.8 min |

**PSNR per scene**

| | garden | bicycle | stump | bonsai | counter | kitchen | room |
|---|---:|---:|---:|---:|---:|---:|---:|
| regular | 27.38 | 25.21 | 26.71 | 32.68 | 29.30 | 31.55 | 32.05 |
| POPSpa | 26.34 | 24.64 | 26.18 | 32.08 | 28.92 | 30.95 | 31.76 |
| ADMM | 26.47 | 24.77 | 26.39 | 32.43 | 28.89 | 31.18 | 32.06 |

**LPIPS per scene**

| | garden | bicycle | stump | bonsai | counter | kitchen | room |
|---|---:|---:|---:|---:|---:|---:|---:|
| regular | 0.130 | 0.221 | 0.209 | 0.187 | 0.180 | 0.120 | 0.198 |
| POPSpa | 0.245 | 0.310 | 0.279 | 0.206 | 0.201 | 0.143 | 0.217 |
| ADMM | 0.223 | 0.295 | 0.265 | 0.207 | 0.207 | 0.155 | 0.218 |

In short: POPSpa keeps a quarter of the Gaussians and costs 0.57 dB, 0.025 SSIM and 0.051 LPIPS on average, plus about 20% more time. Outdoor scenes pay the most (garden loses 1 dB); indoor scenes lose 0.3 to 0.6 dB. Renders look a little softer in grass and foliage, structure is intact. ADMM with its defaults keeps more Gaussians (400k) and lands slightly better on all three metrics, so at equal target counts the two methods look close. The POPSpa run's 30k dense evaluation matches the regular run, so the sparsification phase is the only difference. Peak VRAM is a bit lower than the dense run.

**Functional checks**

- `popspa_tests`: 129 passed, 1 skipped. Python panel/property-view tests: 82 passed.
- One existing test fails on this branch: `CheckpointComponentConfigRoundTripTest.AllConfigsRoundTripInsideLfkp` (tests/test_component_config_serialization.cpp:580) expects checkpoint header version 3 and gets 4. Needs an update in the PR.
- GUI: enabling Sparsity, picking POPSpa, setting Target Gaussians, training, saving and reopening the project all work. A 100k model went 100,000 → 44,721 → 20,000 as expected.
- Headless resume and CLI snapshots work as fixed earlier.

**Things worth changing before merge**

1. POPSpa allocates a second full-size Adam state at the first scoring pass while the strategy's optimizer is still alive (trainer_popspa.cpp, make_popspa_optimizer). That is about 0.4 GB per million Gaussians on top of the dense run, 3.7 GB at 10M, right before the prune that would free memory.
2. The POPSpa step counts (5000 + 5000) and the projection interval are not scaled by `steps_scaler`, but the GUI scales every other step setting with the dataset size. Either scale them or say in the tooltip that they are fixed.
3. Every checkpoint is now written as version 4, also with sparsity off, so older builds cannot read new checkpoints. Consider writing v3 unless POPSpa state is present, or call the change out in the PR text.
4. The effective-rank kernel has no handling for non-finite scales; a NaN row poisons its gradients silently.
5. Small: rho and epsilon show as 0.00000000 in the panel for values like 1e-9, and the non-English locale strings are English copies.

**Comment by lyehe** [comment]
Ran a follow-up benchmark on Windows (RTX 5090), using head `770154fa` plus the local COLMAP calibration correction described below. The correction is not yet committed to this PR. Same corrected binary and LFS libraries for every scene; Garden reuses the previously completed corrected run and the other six scenes were trained fresh.

This run follows the paper's tier-L training recipe more closely: MCMC for 30k iterations with a 2.56M cap, then 5k POPSpa optimize-sparsify steps and 5k recovery to 512k Gaussians. Training uses `images_4` outdoors / `images_2` indoors and `--eval --test-every 8`, with the position learning rate adjusted to the reference camera-radius scale. It still uses LFS's native MCMC/rasterizer/optimizer, so this is a similar-setting reproduction, not identical gsplat training.

**Benchmark, mean over 7 scenes**

| | PSNR | SSIM | LPIPS | Gaussians |
|---|---:|---:|---:|---:|
| Paper POPSpa, same seven scenes | 29.4015 | 0.868085 | 0.200943–0.200968 | 512,000 |
| LFS MCMC + POPSpa, corrected calibration | 29.1313 | 0.870305 | 0.198911 | 512,000 |
| Difference, LFS minus paper | -0.2702 | +0.002220 | -0.002057 to -0.002032 | 0 |

**PSNR per scene**

| | garden | bicycle | stump | bonsai | counter | kitchen | room |
|---|---:|---:|---:|---:|---:|---:|---:|
| Paper POPSpa | 26.4970 | 24.8837 | 26.8893 | 33.0177 | 29.7415 | 32.1900 | 32.5914 |
| LFS MCMC + POPSpa | 26.4549 | 24.9244 | 27.3099 | 32.2499 | 29.4188 | 31.7099 | 31.8515 |
| Difference | -0.0421 | +0.0407 | +0.4206 | -0.7678 | -0.3227 | -0.4801 | -0.7399 |

**LPIPS per scene**

| | garden | bicycle | stump | bonsai | counter | kitchen | room |
|---|---:|---:|---:|---:|---:|---:|---:|
| Paper POPSpa | 0.2120 | 0.2979–0.2980 | 0.2647 | 0.1691–0.1692 | 0.1696 | 0.1148–0.1149 | 0.1784 |
| LFS MCMC + POPSpa | 0.2063 | 0.2859 | 0.2474 | 0.1754 | 0.1737 | 0.1189 | 0.1848 |

The closer training recipe reaches within 0.2702 dB of the published same-scene average, with slightly better mean SSIM and LPIPS. Garden and Bicycle are within 0.043 dB; Stump is higher, while Bonsai and Room remain about 0.75 dB lower. This does not establish an exact paper reproduction or resolve those remaining scene-specific gaps.

**Evaluation and checks**

- These tables use the paper evaluator: original `images` resized with PIL bicubic to maximum long side 1600, clamped floating-point predictions without an 8-bit round trip, TorchMetrics SSIM and VGG LPIPS (`normalize=False`). They are not directly comparable with the native metrics in the earlier MRNF table.
- The evaluator was previously checked against the released Garden 256k checkpoint: measured PSNR 25.798456 versus published 25.798500.
- All seven run configs, final PLY counts/hashes, dense/final native metric stages and published-reference hashes were checked. Every final PLY has 512,000 Gaussians.
- One run per scene, with no seed-variance estimate. Flowers and Treehill are unavailable locally, so the reference mean above is recomputed for the same seven scenes, not the paper's nine-scene Table 1 mean. Timing is omitted because this is not a hardware-matched timing comparison.
- Published values come from the uncompressed `gsplat_popspa_erank`, tier `L` operating points in [the project data](https://fraunhoferhhi.github.io/KISS-GS/rd/rd.json). LPIPS ranges preserve slightly different duplicate published records. Training/evaluation details: [paper](https://arxiv.org/html/2608.26948v1); independent reference pinned to [`5d5212a`](https://github.com/OpsiClear/KissGsVanilla/tree/5d5212a33a45a6668beb822f17b4b2d88e0f5f8a).

**Calibration finding and earlier benchmark**

The local correction preserves original COLMAP intrinsics and scales them using actual image-dimension ratios. Previously, nominal downsample-factor scaling combined with truncated dimensions biased projections for images whose original dimensions do not divide evenly. With byte-identical Garden training configs, correcting this changed paper-protocol PSNR from 25.5094 to 26.4549 dB, versus 26.4970 published.

Separately, all 21 runs of the earlier MRNF 1M benchmark were repeated with this corrected binary. Native seven-scene PSNR means were 29.2694 regular, 28.7126 POPSpa (256k), and 28.8975 ADMM (400k), closely reproducing the earlier report. POPSpa still lost 0.5711 dB against its own dense stages. The calibration issue therefore does not explain away the earlier native compaction loss; the MCMC 2.56M-to-512k paper-setting experiment is a different comparison.

Binary SHA-256: `1939c89204dbcd3c0105ce9f1d0fa3444c29825b5beb7472bd7feb5a2adb94e1`. Published-data SHA-256: `d523c19410ed6160ceb42234e61b9e814577e5b51b088ce6a368610c778dbc50`.

**Comment by NickAcPT** [comment]
From my testing locally, this does lead to great results! Although I've noticed a complete UI freeze during all POPSpa score passes (trainer_popspa.cpp:268) which can cause the app to go "Not Responding" and leading Windows to kill it

**Comment by lyehe** [comment]
@NickAcPT I investigated the scoring-time UI freeze and pushed two shared CUDA/Vulkan integration fixes in 0cb3d00c:

- The viewport's 15 ms arena acquisition timeout previously covered CPU ownership only. Its GPU handoff could still block the GUI thread in `cudaDeviceSynchronize`, including on unrelated streams. Pending handoffs now complete off the GUI thread while retaining the arena reservation; a completed producer event avoids the unrelated-stream wait. A controlled 1,000 ms pending producer previously held the request for about 1,009 ms; it now defers in about 15.2 ms.
- Private-scratch cleanup left non-owning Vulkan aliases pointing to retired buffers. Cleanup now invalidates these views as well as their owners, preserving timeline retirement and external parent ownership.

Validation: Release application and all dependent DLLs/Python extension rebuilt together; 29 focused C++ tests and 9 Python lifetime tests passed. Garden trials on both FastGS and GUT reduced 2.56M Gaussians to 512k with two scoring passes over 161 cameras and shortened 10-step optimization/refinement phases. With a moving camera and Khronos validation active, both final trials completed without validation errors, device loss, capture errors, or window-message timeouts. The local validation checkout also includes the previously noted COLMAP calibration correction; that change is outside this commit. The full repository test suite was not run.

These fixes concern the shared rendering integration, without changing POPSpa's optimization algorithm. I have not reproduced your exact whole-pass freeze, so I cannot yet call that report resolved. Scoring progress still stays on one training step during a pass. Could you retest the updated head with a full rebuild of the application and dependent DLLs?

[Investigation, regression coverage, and remaining limits](https://github.com/lyehe/LichtFeld-Studio/blob/0cb3d00c/eval/popspa-ui-freeze-investigation.md).

**Comment by lyehe** [comment]
Resolved the conflicts with master (`ad12e210`) and pushed merge commit `4afbb1b5`.

The resolution retains POPSpa's deferred gradient capture while using upstream's fused Adam settings helper and mean-step mask handling. It also retains upstream's normalized mask threshold handling, POPSpa's per-step parameters, and Vulkan scratch usage flags during alias cleanup. Added an assertion covering preservation of the usage flags.

Validation: compiled the merged trainer, POPSpa trainer, Adam optimizer, FastGS rasterizer, and VkSplat viewport translation units with MSVC Release settings, using existing generated headers/dependencies in a separate worktree. Formatting, Windows source preflight, error-debt audit, and diff checks passed. Of 18 focused Python viewport checks, 17 passed; `test_vksplat_viewer_does_not_reject_orthographic_requests` fails on its `frame_view.ortho_scale` source assertion, which also fails on unmodified master. No full application link, native runtime suite, or GUI rerun was performed for this merge.

**Comment by MrNeRF** [comment]
Heads up: This will be merged after the next release 0.5.4. Trying to carve that out asap.

**Comment by NickAcPT** [comment]
> I investigated the scoring-time UI freeze and pushed two shared CUDA/Vulkan integration fixes in [0cb3d00](https://github.com/MrNeRF/LichtFeld-Studio/commit/0cb3d00c833a26a6c00073db25f214fa3e922974):
> [...]
> Could you retest the updated head with a full rebuild of the application and dependent DLLs?

@lyehe I can confirm that I no longer get the UI freeze during POPSpa score passes!
